bitbot/reminder.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	var event eventType
+
+	b.DB.Where("ID = ?", id).Take(&event)
+
+	event.People = strings.Replace(event.People, message.Name+" ", "", -1)
+	b.DB.Save(&event)
+
+	feedback := fmt.Sprintf("Removed %s from \"%s\"",
+		message.Name,
+		event.Description)
+
+	return feedback
+}
+
+// Remove an event given by his ID
+func removeeventType(message *hbot.Message) string {
+	id, err := getMessageIDFromString(message.Content)
+	if err != nil {
+		return "Wrong command. format is : !remind remove [ID]"
+	}
+
+	var event eventType
+
+	b.DB.Where("ID = ? AND Author = ?", id, message.Name).Take(&event)
+
+	// Feedback Message construction
+	var feedbackMessage string
+	if event.ID == id {
+		feedbackMessage = fmt.Sprintf("Deleted event %d : %s",
+			event.ID,
+			event.Description)
+
+		// Delete
+		b.DB.Delete(&event)
+	} else {
+		feedbackMessage = "No event you own with that ID"
+	}
+
+	return feedbackMessage
+}
+
+// Lists all the awaiting events in PM
+func listeventTypes(message *hbot.Message, bot *hbot.Bot) string {
+	// Get all the db rows, iterate through them, format them and send them to pm
+	rows, err := b.DB.Model(&eventType{}).Rows()
+	if err != nil {
+		log.Error("Reminder: Couldn't get DB rows", err)
+	}
+
+	var (
+		event                   eventType
+		eventDescriptionMessage string
+	)
+
+	for rows.Next() {
+		err := b.DB.ScanRows(rows, &event)
+		if err != nil {
+			log.Error("Reminder: Couldn't get a db row", err)
+		}
+
+		eventDescriptionMessage = fmt.Sprintf(
+			"%d : [ %s ] at %s. eventType author : %s, in channel %s, with %s",
+			event.ID,
+			event.Description,
+			event.Time.Format(timeFormat),
+			event.Author,
+			event.Channel,
+			event.People)
+		bot.Msg(message.Name, eventDescriptionMessage)
+	}
+
+	return "I've PM'd you the list of awaiting events"
+}
+
+// Parses an event adding message and adds the event
+func addeventType(message *hbot.Message, bot *hbot.Bot) string {
+	// Parsing the message
+	channel := message.To
+	author := message.From
+	msg := strings.Split(message.Content, " ")
+	// Everything in the message content that isn't a command or a time
+	// specification is considered description
+	description := strings.Join(msg[2:len(msg)-2], " ")
+
+	// We take the two last parts of the message (with space as the separator)
+	// and parse them as a time.
+	timeOfeventType, err := time.Parse(timeFormat, strings.Join(msg[len(msg)-2:], " "))
+	if err != nil {
+		return fmt.Sprintf(
+			"Couldn't parse request format is \"!remind add Jitsi Meeting %s\"",
+			timeFormat)
+	}
+
+	// Adding it to the DB
+	event := eventType{
+		Channel:     channel,
+		Author:      author,
+		Description: description,
+		Time:        timeOfeventType,
+		People:      fmt.Sprintf("%s ", author)}
+	b.DB.NewRecord(event)
+	b.DB.Create(&event)
+
+	// Launch a background routine that will HL interested people and clean the DB.
+	eventTimer := time.NewTimer(time.Until(event.Time))
+
+	go func() {
```

Comment:
So there's one goroutine per timer?

Reply:
Yes, I couldn't think of a better way to do this.

It might take quite a lot of memory though, so I'm very much opened to suggestions.

Reply:
The alternative is a single goroutine that checks for all timers, I think. Not any better.

Reply:
If we do one goroutine per timer, we could calculate the time delta between now and the reminder and call a sleep until then. that wouldn't consume much in the way of resources.

Reply:
That's clever, I wonder how much resource it will save.